operator/v1alpha1/types_kueue.go

```diff
@@ -0,0 +1,304 @@
+// after the transformation is done.
+type ResourceTransformation struct {
+	// input is the name of the input resource.
+	// resources are pod spec resources like cpu, memory, gpus
+	// +required
+	Input corev1.ResourceName `json:"input"`
+
+	// strategy specifies if the input resource should be replaced or retained.
+	// retain means that we will keep the original resources and
+	// apply a transformation.
+	// replace means that the original resources will be replaced
+	// after the transformation is done.
+	// +optional
+	Strategy *ResourceTransformationStrategy `json:"strategy,omitempty"`
+
+	// outputs specifies the output resources and quantities per unit of input resource.
+	// An empty Outputs combined with a `Replace` Strategy causes the Input resource to be ignored by Kueue.
+	// +optional
+	Outputs corev1.ResourceList `json:"outputs,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=LessThanOrEqualToFinalShare;LessThanInitialShare
+type PreemptionStrategy string
+
+const (
+	LessThanOrEqualToFinalShare PreemptionStrategy = "LessThanOrEqualToFinalShare"
+	LessThanInitialShare        PreemptionStrategy = "LessThanInitialShare"
+)
+
+type FairSharing struct {
+	// enable indicates whether to enable fair sharing for all cohorts.
+	// this is disabled by default.
+	// +optional
+	Enable EnabledOrDisabled `json:"enable"`
+
+	// preemptionStrategies indicates which constraints should a preemption satisfy.
+	// The preemption algorithm will only use the next strategy in the list if the
+	// incoming workload (preemptor) doesn't fit after using the previous strategies.
+	// Possible values are:
+	// - LessThanOrEqualToFinalShare: Only preempt a workload if the share of the preemptor CQ
+	//   with the preemptor workload is less than or equal to the share of the preemptee CQ
+	//   without the workload to be preempted.
+	//   This strategy might favor preemption of smaller workloads in the preemptee CQ,
+	//   regardless of priority or start time, in an effort to keep the share of the CQ
+	//   as high as possible.
+	// - LessThanInitialShare: Only preempt a workload if the share of the preemptor CQ
+	//   with the incoming workload is strictly less than the share of the preemptee CQ.
+	//   This strategy doesn't depend on the share usage of the workload being preempted.
+	//   As a result, the strategy chooses to preempt workloads with the lowest priority and
+	//   newest start time first.
+	// The default strategy is ["LessThanOrEqualToFinalShare", "LessThanInitialShare"].
+	// +optional
+	// +kubebuilder:validation:MaxItems=2
+	PreemptionStrategies []PreemptionStrategy `json:"preemptionStrategies,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=batch/job;ray.io/rayjob;ray.io/raycluster;jobset.x-k8s.io/jobset;kubeflow.org/mpijob;kubeflow.org/paddlejob;kubeflow.org/pytorchjob;kubeflow.org/tfjob;kubeflow.org/xgboostjob;workload.codeflare.dev/appwrapper;pod;deployment;statefulset;leaderworkerset.x-k8s.io/leaderworkerset
```

Comment:
Do we expect this list to expand over time? Or is this fairly stable in Kueue?

I guess as Kueue starts supporting new integrations the operator will be updated?

Reply:
Correct. Each release has added some support for new integrations.

Reply:
So I can remove enums if you think that would be difficult to keep up with.

Reply:
https://github.com/kubernetes-sigs/kueue/issues/4659

Reply:
Up to you :) This makes the UX great, but comes at the cost of some toil each time you cut a new release